pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java

```diff
@@ -82,14 +103,220 @@ public void setTableConfigFileName(String tableConfigFileName) {
+        LOGGER.error("Upload segment verification failed, count is zero after max wait time {} ms.",
+            DEFAULT_MAX_SLEEP_TIME_MS);
+        return false;
+      }
+      LOGGER.warn("Upload segment verification count is zero, will retry after {} ms.", DEFAULT_SLEEP_INTERVAL_MS);
+      Thread.sleep(DEFAULT_SLEEP_INTERVAL_MS);
+    }
+
+    LOGGER.info("Successfully verified segment {} and its current status is {}.", _segmentName, state);
+    return true;
+  }
+
+  /**
+   * Deletes the segment for the given segment name and table name.
+   * @return true if delete successful, else false.
+   */
+  private boolean deleteSegment() {
+    try {
+      TableConfig tableConfig = JsonUtils.fileToObject(new File(_tableConfigFileName), TableConfig.class);
+      _tableName = tableConfig.getTableName();
+
+      ControllerTest.sendDeleteRequest(ControllerRequestURLBuilder.baseUrl(ClusterDescriptor.CONTROLLER_URL)
+          .forSegmentDelete(_tableName, _segmentName));
+      return verifySegmentDeleted();
+    } catch (Exception e) {
+      LOGGER.error("Request to delete the segment {} for the table {} failed.", _segmentName, _tableName, e);
+      return false;
+    }
+  }
+
+  /**
+   * Verify given table name and segment name deleted from the controller.
+   * @return true if no segment found, else false.
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  private boolean verifySegmentDeleted()
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    while (getCountForSegmentName() > 0) {
+      if ((System.currentTimeMillis() - startTime) > DEFAULT_MAX_SLEEP_TIME_MS) {
+        LOGGER.error("Delete segment verification failed, count is greater than zero after max wait time {} ms.",
+            DEFAULT_MAX_SLEEP_TIME_MS);
+        return false;
+      }
+      LOGGER.warn("Delete segment verification count greater than zero, will retry after {} ms.",
+          DEFAULT_SLEEP_INTERVAL_MS);
+      Thread.sleep(DEFAULT_SLEEP_INTERVAL_MS);
+    }
+
+    LOGGER.info("Successfully delete the segment {} for the table {}.", _segmentName, _tableName);
+    return true;
+  }
+
+  /**
+   * Retrieve external view for the given table name.
+   * @return TableViews.TableView of OFFLINE and REALTIME segments.
+   */
+  private TableViews.TableView getExternalViewForTable()
+      throws IOException {
+    return JsonUtils.stringToObject(ControllerTest.sendGetRequest(
+        ControllerRequestURLBuilder.baseUrl(ClusterDescriptor.CONTROLLER_URL).forTableExternalView(_tableName)),
+        TableViews.TableView.class);
+  }
+
+  /**
+   * Retrieve the number of segments for both OFFLINE and REALTIME which are in state matching the parameter.
+   * @param state of the segment to be verified in the controller.
+   * @return count for OFFLINE and REALTIME segments.
+   */
+  private long getSegmentCountInState(String state)
```

Comment:
The code here will work, but seems to be implying that a segment can exist in both realtime and offline tables with the same name.

You can choose to just consider the offline side for now. Let us add the realtime side later.

Reply:
Implementation checks for the segment in external view in both offline and realtime. Any specific reason why REALTIME should not be considered for this compatibility test? Aren't we planning to ingest the data from Kafka for these use cases?

Reply:
Realtime segments are not pushed in this manner. The are created by Pinot and pushed into deep store.
You have two choices:
(1) Pick up the table config and check that the table type is OFFLINE. If not, fail the operation. If it is OFFLINE, then get the offline externalview. This is the more rigorous way of doing it.
(2) Assume that the operation is specified correctly, and just check the offline side.

When we start adding realtime events and segments get generated, we will move this code to wherever it fits so that we can re-use the method to check for realtime segments being ONLINE or CONSUMING state.

Reply:
ok. Will do with option 2 for now.

Reply:
Fixed